src/database/database.py

```diff
@@ -316,28 +324,159 @@ def get_intolerances(self, user_id: str) -> list[str]:
+
+        The list of valid characters includes:
+        - Uppercase and lowercase Latin letters
+        - Arabic numerals
+        - Underscores, hyphens, periods, and dollar signs
+
+        As a regular expression, a character is valid if it matches the following pattern:
+        `[a-zA-Z0-9_-\.\$]`
+        """
+        for char in username:
+            if not self.is_valid_username_char(char):
+                return False
+        return True
+
+    def add_user(
+        self,
+        id: str,
+        email: str,
+        username: str = None,
+        given_name: str = "",
+        family_name: str = "",
+        profile_image: str = "",
+    ):
         """
         Creates a new user with the specified information and adds it to the database, then returns the created user.
 
         Args:
             id (str): The ID of the user. This value must be unique across all users.
             email (str): The user's email. This field is not checked for validity.
-            name (str): The user's name. This value is optional.
+            username (str): The user's username. This value must be unique across all users and must be at least three characters.
+                If the username is not provided, then a new one will be randomly generated based on the user's given name.
+            given_name (str): The user's given name. This value is optional.
+            family_name (str): The user's family name. This value is optional.
+            profile_image (str): The URL for the user's profile picture. This value is optional.
 
         Returns:
             The newly created `User` object.
 
         Raises:
-            DuplicateUserException: If a user with the specified ID already exists.
+            DuplicateUserException: If a user with the specified ID (or username) already exists.
             DatabaseException: If there was a problem adding the user.
         """
         from database.models import User
 
         if self.user_exists(id):
             raise DuplicateUserException(id)
+        if username != None and self.username_exists(username):
+            raise DuplicateUserException(username)
+
+        # If the username is None, generate a new one
+        if username == None:
+            username = self.generate_username(given_name + " " + family_name)
+
+        # If the username is too short, raise an exception
+        if len(username) < 3:
+            raise DatabaseException(f'The specified username "{username}" is too short')
+
+        # If the username contains invalid characters, raise an exception
+        if not self.username_is_valid(username):
+            raise DatabaseException(
+                f'The specified username "{username}" contains invalid characters'
+            )
 
-        user = User(id=id, email=email, name=name)
+        user = User(
+            id=id,
+            email=email,
+            given_name=given_name,
```

Comment:
You can just have name=name instead of the two scopes that get the first and last name separately

Reply:
No you cant. The "name" field doesnt exist anymore because it has been split into given_name and family_name in the database